cirq/neutral_atoms/neutral_atom_devices.py

```diff
@@ -0,0 +1,424 @@
+                gate
+        """
+        self.validate_operation(operation)
+        if isinstance(operation, (ops.GateOperation,
+                                  ops.ParallelGateOperation)):
+            if isinstance(operation.gate, MeasurementGate):
+                return self._measurement_duration
+        return self._gate_duration
+
+    def validate_gate(self, gate: ops.Gate):
+        """
+        Raises an error if the provided gate isn't part of the native gate set.
+
+        Args:
+            gate: the gate to validate
+
+        Raises:
+            ValueError: If the given gate is not part of the native gate set.
+
+        """
+        if not isinstance(gate, (ops.CCXPowGate,
+                                 ops.CCZPowGate,
+                                 ops.CZPowGate,
+                                 ops.CNotPowGate,
+                                 ops.XPowGate,
+                                 ops.YPowGate,
+                                 ops.PhasedXPowGate,
+                                 MeasurementGate,
+                                 ops.ZPowGate,
+                                 ops.IdentityGate)):
+            raise ValueError('Unsupported gate: {!r}'.format(gate))
+        if isinstance(gate, (ops.CNotPowGate,
+                             ops.CZPowGate,
+                             ops.CCXPowGate,
+                             ops.CCZPowGate)):
+            if not gate.exponent.is_integer():
+                raise ValueError('controlled gates must have integer exponents')
+
+    def validate_operation(self, operation: ops.Operation):
+        """
+        Raises an error if the given operation is invalid on this device.
+
+        Args:
+            operation: the operation to validate
+
+        Raises:
+            ValueError: If the operation is not valid
+        """
+        if not isinstance(operation, (ops.GateOperation,
+                                      ops.ParallelGateOperation)):
+            raise ValueError('Unsupported operation: {!r}'.format(operation))
+
+        # The gate must be valid
+        self.validate_gate(operation.gate)
+
+        # All qubits the operation acts on must be on the device
+        for q in operation.qubits:
+            if q not in self.qubits:
+                raise ValueError('Qubit not on device: {!r}'.format(q))
+
+        if isinstance(operation.gate, (ops.MeasurementGate, ops.IdentityGate)):
+            return
+
+        # Verify that a controlled gate operation is valid
+        if isinstance(operation, ops.GateOperation):
+            if len(operation.qubits) > self._max_parallel_c:
+                raise ValueError("Too many qubits acted on in parallel by a "
+                                 "controlled gate operation")
+            if len(operation.qubits) > 1:
+                for p in operation.qubits:
+                    for q in operation.qubits:
+                        if (isinstance(p, GridQubit) and
```

Comment:
We require that qubits be grid qubits, right? They have to be on the device.

Reply:
I'm not sure if you mean that the GridQubit check is the problem or if the fact that they are GridQubits implies that they are on the device. If it is the first, then I included this check to satisfy mypy. If it is the second, GridQubits are "located" by their indices. If the device only has (0,0), (0,1), (1,0), and (1,1), and you try to act on (2,2), there should be a complaint.

Reply:
The fact that they are on the device implies that they are grid qubits (or rather something that supports distance). Use `cast(..., ...)` from `typing` to convince mypy.

Reply:
I now use cast